fish-rust/src/builtins/echo.rs

```diff
@@ -0,0 +1,228 @@
+/// - `nnn`: octal value, one to three digits
+/// - `xhh`: hex value, one to two digits
+fn parse_numeric_sequence<I>(chars: I) -> Option<(usize, u8)>
+where
+    I: IntoIterator<Item = char>,
+{
+    let mut chars = chars.into_iter().peekable();
+
+    // the first character of the numeric part of the sequence
+    let mut start = 0;
+
+    let mut base: u8 = 0;
+    let mut max_digits = 0;
+
+    let first = *chars.peek()?;
+    if first.is_digit(8) {
+        // Octal escape
+        base = 8;
+
+        // If the first digit is a 0, we allow four digits (including that zero); otherwise, we
+        // allow 3.
+        max_digits = if first == '0' { 4 } else { 3 };
+    } else if first == 'x' {
+        // Hex escape
+        base = 16;
+        max_digits = 2;
+
+        // Skip the x
+        start = 1;
+    };
+
+    if base == 0 {
+        return None;
+    }
+
+    let mut val = 0;
+    let mut consumed = start;
+    for digit in chars
+        .skip(start)
+        .take(max_digits)
+        .map_while(|c| c.to_digit(base.into()))
+    {
+        // base is either 8 or 16, so digit can never be >255
+        let digit = u8::try_from(digit).unwrap();
+
+        val = val * base + digit;
+
+        consumed += 1;
+    }
+
+    // We succeeded if we consumed at least one digit.
+    if consumed > 0 {
+        Some((consumed, val))
+    } else {
+        None
+    }
+}
+
+/// The echo builtin.
+///
+/// Bash only respects `-n` if it's the first argument. We'll do the same. We also support a new,
+/// fish specific, option `-s` to mean "no spaces".
+pub fn echo(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    args: &mut [&wstr],
+) -> Result<(), CommandError> {
+    let (opts, optind) = parse_options(args, parser, streams)?;
+
+    // The special character \c can be used to indicate no more output.
+    let mut output_stopped = false;
+
+    // We buffer output so we can write in one go,
+    // this matters when writing to an fd.
+    let mut out = WString::new();
+    let args_to_echo = &args[optind..];
+    'outer: for (idx, arg) in args_to_echo.iter().enumerate() {
+        if opts.print_spaces && idx > 0 {
+            out.push(' ');
+        }
+
+        let mut chars = arg.chars().peekable();
+        while let Some(c) = chars.next() {
+            if !opts.interpret_special_chars || c != '\\' {
+                // Not an escape.
+                out.push(c);
+                continue;
+            }
+
+            // Most escapes consume one character in addition to the backslash; the numeric
+            // sequences may consume more, while an unrecognized escape sequence consumes none.
+            let Some(next_char) = chars.peek() else {
+                out.push('\\');
+                break;
+            };
+
+            // Most escapes consume one character in addition to the backslash; the numeric
+            // sequences may consume more, while an unrecognized escape sequence consumes none.
```

Comment:
This comment is duplicated. The second occurrence makes more sense.
Instead of the first I'd write something like `// Incomplete escape sequence is echoed verbatim`.